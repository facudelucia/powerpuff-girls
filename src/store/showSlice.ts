import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { Show, Episode } from "../types/showTypes";

export interface ShowState {
  show: Show | null;
  episodes: Episode[];
  selectedEpisode: Episode | null;
  status: "idle" | "loading" | "failed";
}

const initialState: ShowState = {
  show: null,
  episodes: [],
  selectedEpisode: null,
  status: "idle",
};

export const fetchShowDetails = createAsyncThunk(
  "show/fetchShowDetails",
  async () => {
    const response = await axios.get("https://api.tvmaze.com/shows/6771");
    return response.data;
  }
);

export const fetchEpisodes = createAsyncThunk(
  "show/fetchEpisodes",
  async () => {
    const response = await axios.get(
      "https://api.tvmaze.com/shows/6771/episodes"
    );
    return response.data;
  }
);

const showSlice = createSlice({
  name: "show",
  initialState,
  reducers: {
    selectEpisode: (state, action) => {
      state.selectedEpisode =
        state.episodes.find(
          (episode: Episode) => episode.id === action.payload
        ) || null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchShowDetails.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchShowDetails.fulfilled, (state, action) => {
        state.status = "idle";
        state.show = action.payload;
      })
      .addCase(fetchEpisodes.fulfilled, (state, action) => {
        state.episodes = action.payload;
      });
  },
});

export const { selectEpisode } = showSlice.actions;
export default showSlice.reducer;
