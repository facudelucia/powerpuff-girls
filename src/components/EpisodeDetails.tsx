import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../store';
import styled, { keyframes } from 'styled-components';

const Container = styled.div`
  max-width: 800px;
  margin: auto;
  padding: 20px;
  font-family: Arial, sans-serif;

  @media (max-width: 768px) {
    padding: 15px;
  }
`;

const Title = styled.h1`
  font-size: 2rem;
  color: #333;

  @media (max-width: 768px) {
    font-size: 1.5rem;
  }
`;

const Image = styled.img`
  max-width: 100%;
  height: auto;
  margin-bottom: 20px;
  border-radius: 8px;
  
  @media (max-width: 768px) {
    margin-bottom: 15px;
  }
`;

const Summary = styled.div`
  font-size: 1rem;
  color: #555;
  line-height: 1.6;
`;

const LoadingSpinner = styled.div`
  border: 4px solid #f3f3f3;
  border-radius: 50%;
  border-top: 4px solid #3498db;
  width: 40px;
  height: 40px;
  animation: ${keyframes`
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
  `} 2s linear infinite;
  margin: 20px auto;
`;

const EpisodeDetails: React.FC = () => {
  const selectedEpisode = useSelector((state: RootState) => state.show.selectedEpisode);

  if (!selectedEpisode) return <LoadingSpinner data-testid="loading"/>;

  return (
    <Container>
      <Title>{selectedEpisode.name}</Title>
      {selectedEpisode.image && (
        <Image src={selectedEpisode.image.original} alt={selectedEpisode.name} />
      )}
      <Summary dangerouslySetInnerHTML={{ __html: selectedEpisode.summary }} />
    </Container>
  );
};

export default EpisodeDetails;
