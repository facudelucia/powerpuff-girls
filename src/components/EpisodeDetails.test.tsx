import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../store';
import EpisodeDetails from './EpisodeDetails';
import renderer from 'react-test-renderer';

describe('EpisodeDetails Component', () => {
  test('renders loading text before data is loaded', async () => {
    render(
      <Provider store={store}>
        <EpisodeDetails />
      </Provider>
    );

    await waitFor(() => {
      expect(screen.getByTestId('loading')).toBeInTheDocument();
    });
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <EpisodeDetails />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
