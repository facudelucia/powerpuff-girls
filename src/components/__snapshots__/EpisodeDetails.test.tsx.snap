// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeDetails Component matches snapshot 1`] = `
<div
  className="sc-dsLQwm hzXjGX"
  data-testid="loading"
/>
`;
